<h3>Question (ID-3012488):</h3><h2>What is the python "with" statement designed for?</h2><p>I came across the Python <code>with</code> statement for the first time today.  I've been using Python lightly for several months and didn't even know of its existence!  Given its somewhat obscure status, I thought it would be worth asking:</p>

<ol>
<li>What is the Python <code>with</code> statement
designed to be used for?   </li>
<li>What do
you use it for? </li>
<li>Are their any
gotchas I need to be aware of, or
common anti-patterns associated with
its use?  Any cases where it is better use <code>try..finally</code> than <code>with</code>?</li>
<li>Why isn't it used more widely?</li>
<li>Which standard library classes are compatible with it?</li>
</ol>
<br /><h3>Answers (Total-8):</h3><b>#0</b><br /><p>I would suggest two interesting lectures:  </p>

<ul>
<li><a href="http://www.python.org/dev/peps/pep-0343/" rel="nofollow">PEP 343</a> The "with" Statement  </li>
<li><a href="http://effbot.org/zone/python-with-statement.htm" rel="nofollow">Effbot</a> Understanding Python's
"with" statement</li>
</ul>

<p><strong>1.</strong>
The <code>with</code> statement is used to wrap the execution of a block with methods defined by a context manager. This allows common <code>try...except...finally</code> usage patterns to be encapsulated for convenient reuse.</p>

<p><strong>2.</strong>
You could do something like:</p>

<pre><code>with open("foo.txt") as foo_file:
    data = foo_file.read()
</code></pre>

<p>OR</p>

<pre><code>from contextlib import nested
with nested(A(), B(), C()) as (X, Y, Z):
   do_something()
</code></pre>

<p>OR (Python 3.1)</p>

<pre><code>with open('data') as input_file, open('result', 'w') as output_file:
   for line in input_file:
     output_file.write(parse(line))
</code></pre>

<p>OR</p>

<pre><code>lock = threading.Lock()
with lock:
    # Critical section of code
</code></pre>

<p><strong>3.</strong>
 I don't see any Antipattern here.<br>
Quoting <a href="http://diveintopython3.org/" rel="nofollow">Dive into Python</a>: </p>

<blockquote>
  <p><strong>try..finally is good. with is better.</strong></p>
</blockquote>

<p><strong>4.</strong>
I guess it's related to programmers's habit to use <code>try..catch..finally</code> statement from other languages.</p>
<br /><b>#1</b><br /><ol>
<li><p>I believe this has already been answered by other users before me, so I only add it for the sake of completeness: the <code>with</code> statement simplifies exception handling by encapsulating common preparation and cleanup tasks in so-called <a href="http://docs.python.org/release/2.5.2/lib/typecontextmanager.html" rel="nofollow">context managers</a>. More details can be found in <a href="http://www.python.org/dev/peps/pep-0343/" rel="nofollow">PEP 343</a>. For instance, the <code>open</code> statement is a context manager in itself, which lets you open a file, keep it open as long as the execution is in the context of the <code>with</code> statement where you used it, and close it as soon as you leave the context, no matter whether you have left it because of an exception or during regular control flow. The <code>with</code> statement can thus be used in ways similar to the <a href="http://en.wikipedia.org/wiki/Resource_Acquisition_Is_Initialization" rel="nofollow">RAII pattern</a> in C++: some resource is acquired by the <code>with</code> statement and released when you leave the <code>with</code> context.</p></li>
<li><p>Some examples are: opening files using <code>with open(filename) as fp:</code>, acquiring locks using <code>with lock:</code> (where <code>lock</code> is an instance of <code>threading.Lock</code>). You can also construct your own context managers using the <code>contextmanager</code> decorator from <code>contextlib</code>. For instance, I often use this when I have to change the current directory temporarily and then return to where I was:</p>

<pre><code>from contextlib import contextmanager
import os


@contextmanager
def working_directory(path):
    current_dir = os.getcwd()
    os.chdir(path)
    try:
        yield
    finally:
        os.chdir(current_dir)


with working_directory("data/stuff"):
    # do something within data/stuff
# here I am back again in the original working directory
</code></pre>

<p>Here's another example that temporarily redirects <code>sys.stdin</code>, <code>sys.stdout</code> and <code>sys.stderr</code> to some other file handle and restores them later:</p>

<pre><code>from contextlib import contextmanager
import sys


@contextmanager
def redirected(**kwds):
    stream_names = ["stdin", "stdout", "stderr"]
    old_streams = {}
    try:
        for sname in stream_names:
            stream = kwds.get(sname, None)
            if stream is not None and stream != getattr(sys, sname):
                old_streams[sname] = getattr(sys, sname)
                setattr(sys, sname, stream)
        yield
    finally:
        for sname, stream in old_streams.iteritems():
            setattr(sys, sname, stream)


with redirected(stdout=open("/tmp/log.txt", "w")):
     # these print statements will go to /tmp/log.txt
     print "Test entry 1"
     print "Test entry 2"
# back to the normal stdout
print "Back to normal stdout again"
</code></pre>

<p>And finally, another example that creates a temporary folder and cleans it up when leaving the context:</p>

<pre><code>from tempfile import mkdtemp
from shutil import rmtree


@contextmanager
def temporary_dir(*args, **kwds):
    name = mkdtemp(*args, **kwds)
    try:
        yield name
    finally:
        shutil.rmtree(name)


with temporary_dir() as dirname:
    # do whatever you want
</code></pre></li>
</ol>
<br /><b>#2</b><br /><p>The Python <code>with</code> statement is built-in language support of the <a href="http://en.wikipedia.org/wiki/Resource_Acquisition_Is_Initialization" rel="nofollow"><code>Resource Acquisition Is Initialization</code></a> idiom commonly used in C++. It is intended to allow safe acquisition and release of operating system resources.</p>

<p>The <code>with</code> statement creates resources within a scope/block. You write your code using the resources within the block. When the block exits the resources are cleanly released regardless of the outcome of the code in the block (that is whether the block exists normally or because of an exception).</p>

<p>Many resources in the Python library that obey the protocol required by the <code>with</code> statement and so can used with it out-of-the-box. However anyone can make resources that can be used in a with statement by implementing the well documented protocol: <a href="http://www.python.org/dev/peps/pep-0343/" rel="nofollow">PEP 0343</a></p>

<p>Use it when ever you acquire resources in your application that must be explicitly relinquished such as files, network connections, locks and the like.</p>
<br /><b>#3</b><br /><p>See <a href="http://www.python.org/dev/peps/pep-0343/" rel="nofollow">PEP 343 - The 'with' statement</a>, there is an example section at the end.</p>

<blockquote>
  <p>... new statement "with" to the Python
  language to make
      it possible to factor out standard uses of try/finally statements.</p>
</blockquote>
<br /><b>#4</b><br /><p>An example of an antipattern might be to use the <code>with</code> inside a loop when it would be more efficient to have the <code>with</code> outside the loop</p>

<p>for example</p>

<pre><code>for row in lines:
    with open("outfile","a") as f:
        f.write(row)
</code></pre>

<p>vs</p>

<pre><code>with open("outfile","a") as f:
    for row in lines:
        f.write(row)
</code></pre>

<p>The first way is opening and closing the file for each <code>row</code> which may cause performance problems compared to the second way with opens and closes the file just once.</p>
<br /><b>#5</b><br /><p>Again for completeness I'll add my most useful use-case for <code>with</code> statements.</p>

<p>I do a lot of scientific computing and for some activities I need the <code>Decimal</code> library for arbitrary precision calculations.  Some part of my code I need high precision and for most other parts I need less precision.</p>

<p>I set my default precision to a low number and then use <code>with</code> to get a more precise answer for some sections:</p>

<pre><code>from decimal import localcontext

with localcontext() as ctx:
    ctx.prec = 42   # Perform a high precision calculation
    s = calculate_something()
s = +s  # Round the final result back to the default precision
</code></pre>

<p>I use this a lot with the Hypergeometric Test which requires the division of large numbers resulting form factorials.  When you do genomic scale calculations you have to be careful of round-off and overflow errors.</p>
<br /><b>#6</b><br /><p>The with statement works with so-called context managers:</p>

<p><a href="http://docs.python.org/release/2.5.2/lib/typecontextmanager.html" rel="nofollow">http://docs.python.org/release/2.5.2/lib/typecontextmanager.html</a></p>

<p>The idea is to simplify exception handling by doing the necessary cleanup after leaving the 'with' block. Some of the python built-ins already work as context managers.</p>
<br /><b>#7</b><br /><p>points 1, 2, and 3 being reasonably well covered:</p>

<p><strong>4:</strong> it is relatively new, only available in python2.6+ (or python2.5 using <code>from __future__ import with_statement</code>)</p>
<br />